In [1]:
from sklearn.model_selection import train_test_split
import pandas as pd
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import OneHotEncoder, OrdinalEncoder, StandardScaler
from sklearn.compose import ColumnTransformer, make_column_transformer
import numpy as np
from sklearn.svm import SVC, SVR
from sklearn.metrics import (
    accuracy_score,
    classification_report,
    confusion_matrix,
    f1_score,
    make_scorer,
    precision_score,
    recall_score,
    average_precision_score, 
    auc
)
from sklearn.model_selection import (
    GridSearchCV,
    RandomizedSearchCV,
    cross_val_score,
    cross_validate,
    train_test_split,
)
from sklearn.linear_model import LogisticRegression, Ridge
from sklearn.ensemble import RandomForestClassifier, RandomForestRegressor

from sklearn.multiclass import OneVsOneClassifier, OneVsRestClassifier
from sklearn.feature_selection import RFE
from sklearn.feature_selection import SelectFromModel

In [3]:
white_df = pd.read_csv("../data/raw/winequality-white.csv", sep=";")
red_df = pd.read_csv("../data/raw/winequality-red.csv", sep=";")

red_df['type']='red wine'
white_df['type']='white wine'
wine = pd.concat([red_df,white_df]).reset_index().drop(columns = ['index'])

We combined both csv files together and created a "type" column to see whether the type of the wine will affect the judgement of the quality. Because sometimes, human's perception of wine type may affect the independent scoring on the wine quality, so that's why we added a binary feature to account for this factor. 

In [4]:
train_df, test_df = train_test_split(wine, test_size=0.2, random_state=123)

In [5]:
X_train = train_df.drop(columns=["quality"])
X_test = test_df.drop(columns=["quality"])

y_train = train_df["quality"]
y_test = test_df["quality"]

In [6]:
numeric_feats = X_train.select_dtypes(include=[np.number]).columns.values.tolist()
binary_feats = ["type"]

numeric_transformer = make_pipeline(StandardScaler())
binary_transformer = make_pipeline(OneHotEncoder(drop="if_binary", dtype=int))

preprocessor = make_column_transformer(
    (numeric_transformer, numeric_feats),
    (binary_transformer, binary_feats)
)

In our data file, we only have numerical features originally; however we need to scale those number before we feed them in the model. And we engineered a binary feature ("type"), we used `OneHotEncoder` with `drop="if_binary"` argument to form the transformer. 

In [7]:
column_name = numeric_feats + binary_feats
pd.DataFrame(preprocessor.fit_transform(X_train, y_train), columns = column_name)

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0         -1.552156         -0.537847    -0.616843       -0.849603  -0.256654   
1         -0.551023         -0.903607    -0.274431       -0.849603  -0.285082   
2          0.065059         -0.720727     0.547359        1.949924  -0.398798   
3         -0.165971         -0.598807     0.273429       -0.556020  -0.626230   
4         -0.011951         -0.598807    -0.137466       -0.807663  -0.228225   
...             ...               ...          ...             ...        ...   
5192      -0.319992          0.681350    -0.274431        4.319561  -0.711516   
5193       0.758151         -0.476888     0.067982       -0.597961  -0.086080   
5194      -0.859063          1.534788    -2.123458       -0.702812  -0.000793   
5195       0.604131         -0.720727    -0.274431        1.792647  -0.086080   
5196       0.296090         -0.842647    -0.137466        1.834587  -0.000793   

      free sulfur dioxide  total sulfur dioxide   density        pH  \
0               -0.599698             -0.688718 -1.550251  0.637496   
1               -0.487984             -0.460217 -1.427678 -0.236615   
2                0.768801              0.234073  1.692956  0.887242   
3               -0.487984             -0.073524 -1.553564 -0.174178   
4               -0.208698              0.260438 -0.460348  0.200441   
...                   ...                   ...       ...       ...   
5192            -0.208698              0.102245  2.736480 -0.985853   
5193            -1.102412             -0.794180 -0.221828 -2.047274   
5194            -1.437555             -1.813644  0.010066  1.886227   
5195             2.919299              1.420518  1.129784 -0.486361   
5196             1.243586              1.692961  1.427934 -0.486361   

      sulphates   alcohol  type  
0      0.602356  0.755569   1.0  
1     -0.134424  1.005574   1.0  
2     -0.804224 -1.161136   1.0  
3     -1.005164  1.755589   1.0  
4     -0.536304  0.005554   1.0  
...         ...       ...   ...  
5192  -0.737244  0.088889   1.0  
5193  -0.268384 -0.827796   1.0  
5194   0.200476  0.755569   0.0  
5195  -0.536304 -0.577791   1.0  
5196  -0.469324 -1.327806   1.0  

[5197 rows x 12 columns]

In [8]:
# Imported from DSCI 573 Lecture Notes from UBC
def mean_std_cross_val_scores(model, X_train, y_train, **kwargs):
    """
    Returns mean and std of cross validation

    Parameters
    ----------
    model :
        scikit-learn model
    X_train : numpy array or pandas DataFrame
        X in the training data
    y_train :
        y in the training data

    Returns
    ----------
        pandas Series with mean scores from cross_validation
    """

    scores = cross_validate(model, X_train, y_train, **kwargs)

    mean_scores = pd.DataFrame(scores).mean()
    std_scores = pd.DataFrame(scores).std()
    out_col = []

    for i in range(len(mean_scores)):
        out_col.append((f"%0.3f (+/- %0.3f)" % (mean_scores[i], std_scores[i])))

    return pd.Series(data=out_col, index=mean_scores.index)

In [19]:
# Imported from DSCI 573 Lecture Notes from UBC
def mape(true, pred):
    return 100.0 * np.mean(np.abs((pred - true) / true))


# make a scorer function that we can pass into cross-validation
mape_scorer = make_scorer(mape, greater_is_better=False)

score_types_reg = {
    "neg_mean_squared_error": "neg_mean_squared_error",
    "neg_root_mean_squared_error": "neg_root_mean_squared_error",
    "neg_mean_absolute_error": "neg_mean_absolute_error",
    "r2": "r2",
    "mape_scorer": mape_scorer,
}

models = {
    "Ridge": Ridge(max_iter=2000),
    "SVR": SVR(),
    "OneVsRest":OneVsRestClassifier(LogisticRegression()),
    "Random Forest": RandomForestRegressor(random_state=123)
}

In [28]:
import imblearn
from imblearn.pipeline import make_pipeline as make_imb_pipeline
from imblearn.over_sampling import RandomOverSampler

In [36]:
results_comb={}
for keys in models.keys():
    pipe_comb = make_imb_pipeline(RandomOverSampler(sampling_strategy="minority"), preprocessor, models[keys])
    results_comb[keys]=mean_std_cross_val_scores(
        pipe_comb, X_train, y_train, return_train_score=True, scoring=score_types_reg
    )
pd.DataFrame(results_comb)

Ridge                 SVR  \
fit_time                             0.023 (+/- 0.004)   1.073 (+/- 0.064)   
score_time                           0.005 (+/- 0.001)   0.350 (+/- 0.007)   
test_neg_mean_squared_error         -1.093 (+/- 0.097)  -0.503 (+/- 0.018)   
train_neg_mean_squared_error        -1.071 (+/- 0.048)  -0.427 (+/- 0.013)   
test_neg_root_mean_squared_error    -1.044 (+/- 0.047)  -0.709 (+/- 0.013)   
train_neg_root_mean_squared_error   -1.035 (+/- 0.023)  -0.654 (+/- 0.010)   
test_neg_mean_absolute_error        -0.806 (+/- 0.025)  -0.531 (+/- 0.009)   
train_neg_mean_absolute_error       -0.802 (+/- 0.022)  -0.474 (+/- 0.004)   
test_r2                             -0.436 (+/- 0.141)   0.339 (+/- 0.019)   
train_r2                            -0.405 (+/- 0.063)   0.440 (+/- 0.016)   
test_mape_scorer                   -14.606 (+/- 0.662)  -9.431 (+/- 0.282)   
train_mape_scorer                  -14.539 (+/- 0.449)  -8.405 (+/- 0.096)   

                                             OneVsRest       Random Forest  
fit_time                             0.143 (+/- 0.009)   1.616 (+/- 0.016)  
score_time                           0.005 (+/- 0.000)   0.021 (+/- 0.000)  
test_neg_mean_squared_error         -0.880 (+/- 0.072)  -0.402 (+/- 0.030)  
train_neg_mean_squared_error        -0.835 (+/- 0.046)  -0.057 (+/- 0.001)  
test_neg_root_mean_squared_error    -0.938 (+/- 0.039)  -0.634 (+/- 0.024)  
train_neg_root_mean_squared_error   -0.913 (+/- 0.025)  -0.239 (+/- 0.003)  
test_neg_mean_absolute_error        -0.590 (+/- 0.017)  -0.457 (+/- 0.015)  
train_neg_mean_absolute_error       -0.571 (+/- 0.013)  -0.170 (+/- 0.001)  
test_r2                             -0.154 (+/- 0.091)   0.473 (+/- 0.027)  
train_r2                            -0.095 (+/- 0.061)   0.925 (+/- 0.002)  
test_mape_scorer                   -10.269 (+/- 0.324)  -8.211 (+/- 0.405)  
train_mape_scorer                   -9.918 (+/- 0.234)  -3.071 (+/- 0.035)

After comparing different regression models by using various matrix, we found the best model is Random Forest, because we got highest cross-validation score. However, we figured out that we may encounter some overfitting issue with Random Forest model as the difference between train score and validation score is quite wide. So, we further conduct feature selections and hyper-parameter optimization as follows. 

In [37]:
rfe = RFE(RandomForestRegressor(random_state=123), n_features_to_select=10)

pipe_rf_rfe = make_imb_pipeline(RandomOverSampler(sampling_strategy="minority"), preprocessor, rfe, RandomForestRegressor(random_state=123))

results_comb['Random Forest_rfe'] = mean_std_cross_val_scores(pipe_rf_rfe, X_train, y_train, return_train_score=True, scoring=score_types_reg)

In [38]:
pd.DataFrame(results_comb)

Ridge                 SVR  \
fit_time                             0.023 (+/- 0.004)   1.073 (+/- 0.064)   
score_time                           0.005 (+/- 0.001)   0.350 (+/- 0.007)   
test_neg_mean_squared_error         -1.093 (+/- 0.097)  -0.503 (+/- 0.018)   
train_neg_mean_squared_error        -1.071 (+/- 0.048)  -0.427 (+/- 0.013)   
test_neg_root_mean_squared_error    -1.044 (+/- 0.047)  -0.709 (+/- 0.013)   
train_neg_root_mean_squared_error   -1.035 (+/- 0.023)  -0.654 (+/- 0.010)   
test_neg_mean_absolute_error        -0.806 (+/- 0.025)  -0.531 (+/- 0.009)   
train_neg_mean_absolute_error       -0.802 (+/- 0.022)  -0.474 (+/- 0.004)   
test_r2                             -0.436 (+/- 0.141)   0.339 (+/- 0.019)   
train_r2                            -0.405 (+/- 0.063)   0.440 (+/- 0.016)   
test_mape_scorer                   -14.606 (+/- 0.662)  -9.431 (+/- 0.282)   
train_mape_scorer                  -14.539 (+/- 0.449)  -8.405 (+/- 0.096)   

                                             OneVsRest       Random Forest  \
fit_time                             0.143 (+/- 0.009)   1.616 (+/- 0.016)   
score_time                           0.005 (+/- 0.000)   0.021 (+/- 0.000)   
test_neg_mean_squared_error         -0.880 (+/- 0.072)  -0.402 (+/- 0.030)   
train_neg_mean_squared_error        -0.835 (+/- 0.046)  -0.057 (+/- 0.001)   
test_neg_root_mean_squared_error    -0.938 (+/- 0.039)  -0.634 (+/- 0.024)   
train_neg_root_mean_squared_error   -0.913 (+/- 0.025)  -0.239 (+/- 0.003)   
test_neg_mean_absolute_error        -0.590 (+/- 0.017)  -0.457 (+/- 0.015)   
train_neg_mean_absolute_error       -0.571 (+/- 0.013)  -0.170 (+/- 0.001)   
test_r2                             -0.154 (+/- 0.091)   0.473 (+/- 0.027)   
train_r2                            -0.095 (+/- 0.061)   0.925 (+/- 0.002)   
test_mape_scorer                   -10.269 (+/- 0.324)  -8.211 (+/- 0.405)   
train_mape_scorer                   -9.918 (+/- 0.234)  -3.071 (+/- 0.035)   

                                    Random Forest_rfe  
fit_time                            6.043 (+/- 0.045)  
score_time                          0.021 (+/- 0.000)  
test_neg_mean_squared_error        -0.403 (+/- 0.030)  
train_neg_mean_squared_error       -0.057 (+/- 0.001)  
test_neg_root_mean_squared_error   -0.634 (+/- 0.024)  
train_neg_root_mean_squared_error  -0.239 (+/- 0.003)  
test_neg_mean_absolute_error       -0.457 (+/- 0.015)  
train_neg_mean_absolute_error      -0.170 (+/- 0.001)  
test_r2                             0.472 (+/- 0.029)  
train_r2                            0.925 (+/- 0.002)  
test_mape_scorer                   -8.220 (+/- 0.417)  
train_mape_scorer                  -3.074 (+/- 0.032)

In [39]:
pipe_rf_rfe.fit(X_train, y_train)
pipe_rf_rfe.named_steps['rfe'].ranking_

array([1, 1, 1, 1, 1, 1, 1, 2, 1, 1, 1, 3])

In [40]:
rfe_fs = pipe_rf_rfe.named_steps["rfe"].support_
rfe_fs

array([ True,  True,  True,  True,  True,  True,  True, False,  True,
        True,  True, False])

In [41]:
rfe_selected_feats = X_train.columns[rfe_fs]
rfe_selected_feats

Index(['fixed acidity', 'volatile acidity', 'citric acid', 'residual sugar',
       'chlorides', 'free sulfur dioxide', 'total sulfur dioxide', 'pH',
       'sulphates', 'alcohol'],
      dtype='object')

By using Recursive Feature Elimination algorithm, we chose to drop "type" and "density" features. Because by dropping those two features, even though we did not get much better scores, we are able to achieve the same scores with lesser features. It is great because it will simplify our model and save cost for future data collection. 

In [42]:
from scipy.stats import randint

In [43]:
param_dist = {"randomforestregressor__max_depth": randint(low=5, high=1000),
             "randomforestregressor__max_leaf_nodes": randint(low=5, high=1000),
             "randomforestregressor__n_estimators": randint(low=5, high=1000),}

random_search = RandomizedSearchCV(
    pipe_rf_rfe,
    param_distributions=param_dist,
    n_jobs=-1,
    n_iter=10,
    cv=5,
    random_state=123
)

random_search.fit(X_train, y_train)

RandomizedSearchCV(cv=5,
                   estimator=Pipeline(steps=[('randomoversampler',
                                              RandomOverSampler(sampling_strategy='minority')),
                                             ('columntransformer',
                                              ColumnTransformer(transformers=[('pipeline-1',
                                                                               Pipeline(steps=[('standardscaler',
                                                                                                StandardScaler())]),
                                                                               ['fixed '
                                                                                'acidity',
                                                                                'volatile '
                                                                                'acidity',
                                                          

In [44]:
random_search.best_params_

{'randomforestregressor__max_depth': 344,
 'randomforestregressor__max_leaf_nodes': 851,
 'randomforestregressor__n_estimators': 258}

In [45]:
random_search.best_score_

0.4756738804987327

In [46]:
random_search.score(X_train, y_train)

0.9087659489993155

In [47]:
random_search.score(X_test, y_test)

0.5225714417802303

|  | Scores |
| -----------: | --------------: |
|          Best Cross-validation Score |         0.475673    |
|          Best Train Score |            0.908765 |
|          Best Test Score  |             0.522571 |

Finally, we conduct hyper-parameter optimization. We employed random search to look for same key hyper-parameters, such as max_depth, max_leaf_nodes, and n_estimators. The best hyper-parameter value we got from the algorithm are ['max_depth': 344, 'max_leaf_nodes': 851, 'n_estimators': 258]. And the best validation score we got is 0.48. And the test score is 0.52 after tunning the hyper-parameters. However, as we discovered above, the train score is 0.91, which means we still have overfitting issue by using Random Forest model. We may investigate further and figure out the next steps later. 